components/swim/src/error.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2016 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use protobuf;
+
+use std::error;
+use std::fmt;
+use std::io;
+use std::result;
+
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum Error {
+    BadMessage(String),
+    IO(io::Error),
+    ProtobufError(protobuf::ProtobufError),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match *self {
+            Error::BadMessage(ref err) => format!("Bad Message: {:?}", err),
+            Error::IO(ref err) => format!("IO Error: {}", err),
+            Error::ProtobufError(ref err) => format!("ProtoBuf Error: {}", err),
+        };
+        write!(f, "{}", msg)
+    }
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Error::BadMessage(ref _err) => "Bad Protobuf Message; should be Ping/Ack/PingReq",
+            Error::IO(ref err) => err.description(),
+            Error::ProtobufError(ref err) => err.description(),
+        }
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Error {
+        Error::IO(err)
```

Comment:
What I had in mind to solve this problem in the other crates would be to do something along these lines:

```rust
match func_might_err() {
  Ok(val) => val,
  Err(io::Error) => {
    msg = "contextual reason for failure";
    Error::ContextualReason(msg)
  },
  Err(e) => Error::from(e),
}
```